### PR TITLE
TRUNK-6034:Introduce a new Abstract ServiceOrder Class

### PR DESCRIPTION
**Pull request by tendomart**
## Description of what I changed

I Introduced a new abstract class ServiceOrder

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6034

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+package org.openmrs;
+
+/**
+ * Provides properties for several order types like TestOrder , ReferralOrder
+ * and others to accommodate referrals depending on the openmrs implementation
+ * use case
+ * 
+ * @since 2.5.0
+ */
+public abstract class ServiceOrder extends Order {
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/ServiceOrder.java:19]: thanks martin,is this class intended to be used some where,and if so we can also add the code where it is used

Comment (tendomart) [reply api/src/main/java/org/openmrs/ServiceOrder.java:19]: True ,thanks @HerbertYiga  there's a sub-task following this .If we get it right here . the rest will be history.

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	 * @since 2.5.0
+	 */
+	public void setLocation(Location location) {this.location = location;}
+
+	/**
+	 * Creates a discontinuation order for this.
+	 *
+	 * @see org.openmrs.Order#cloneForDiscontinuing()
+	 * @return the newly created order
+	 * @since 1.10
```
[the remark this answers is not in the document]
Comment (tendomart) [reply api/src/main/java/org/openmrs/ServiceOrder.java:172]: Thanks for the insight.I actually found myself between a rock and a hard place as to which versioning I should give to the existing methods from 1.10 yet introduced in a totally new class whose version is 2.5.0 .

**Comment by dkayiwa** [comment]
We also have some compiler errors.

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs;
+
+/**
+ * Provides properties for several order types like TestOrder , ReferralOrder
+ * and others to accommodate referrals depending on the openmrs implementation
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/ServiceOrder.java:14]: `to accommodate referrals` sounds like you are using service order for only referalls.

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	public Integer getNumberOfRepeats() {
+		return numberOfRepeats;
+	}
+
+	/**
+	 * Sets numberOfRepeats of a particular order
+	 *
+	 * @param numberOfRepeats to set
+	 */
+	public void setNumberOfRepeats(Integer numberOfRepeats) {this.numberOfRepeats = numberOfRepeats;}
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/ServiceOrder.java:135]: Why are you doing the above on one line?

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	 */
+	@Override
+	public Order cloneForRevision() {
+		return cloneForRevisionHelper(new Order());
+	}
+
+	/**
+	 * @see Order#cloneForRevisionHelper(Order)
+	 */
+	protected abstract Order cloneForRevisionHelper(Order target);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/ServiceOrder.java:183]: Did you intentionally leave the above method empty?

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	 * Default Constructor
+	 */
+	public ServiceOrder() {}
+
+	/**
+	 * @see org.openmrs.Order#copy()
+	 * <strong>Should</strong> copy all sub-class order fields
+	 */
+	@Override
+	public Order copy() {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/ServiceOrder.java:51]: How is this different from the one in the base class?

File: api/src/main/java/org/openmrs/ServiceOrder.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		return copyHelper(new Order());
+	}
+
+	protected abstract Order copyHelper(Order target);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/ServiceOrder.java:55]: Why did you leave this one empty?

File: api/src/main/java/org/openmrs/ServiceOrder.java (hunk 5 above)
Comment (tendomart) [reply api/src/main/java/org/openmrs/ServiceOrder.java:183]: Yes , I was thinking we could carry out the implementation in the child classes ,due to the conflicts encountered.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/ServiceOrder.java:183]: Which conflicts were those?

File: api/src/main/java/org/openmrs/ServiceOrder.java (hunk 7 above)
Comment (tendomart) [reply api/src/main/java/org/openmrs/ServiceOrder.java:55]: If I turn```
 /**
	 * @see Order#cloneForRevisionHelper(Order)
	 */
	protected TestOrder cloneForRevisionHelper(TestOrder target) {
		super.cloneForRevisionHelper(target);
		target.setSpecimenSource(getSpecimenSource());
		target.setLaterality(getLaterality());
		target.setClinicalHistory(getClinicalHistory());
		target.setFrequency(getFrequency());
		target.setNumberOfRepeats(getNumberOfRepeats());
		return target;
	}
```
to this

```
	protected Order cloneForRevisionHelper(Order target) {
		super.cloneForRevisionHelper(target);
		target.setSpecimenSource(getSpecimenSource());
		target.setLaterality(getLaterality());
		target.setClinicalHistory(getClinicalHistory());
		target.setFrequency(getFrequency());
		target.setNumberOfRepeats(getNumberOfRepeats());
		return target;
	}
```

We cannot get the properties of `TestOrder` , as it's intended to sub-class `ServiceOrder`.

this works , but i wonder wether we can pass **ServiceOrder** as a param


	```
protected Order cloneForRevisionHelper(ServiceOrder target) {
		super.cloneForRevisionHelper(target);
		target.setSpecimenSource(getSpecimenSource());
		target.setLaterality(getLaterality());
		target.setClinicalHistory(getClinicalHistory());
		target.setFrequency(getFrequency());
		target.setNumberOfRepeats(getNumberOfRepeats());
		return target;
	}
```

**Comment by dkayiwa** [comment]
Can we also have a `cloneForDiscontinuingHelper`?

**Pull request merged**